events/importer/base.py

```diff
@@ -74,23 +79,83 @@ def _set_multiscript_field(string, event, languages, field):
             if string:
                 event[field][script] = string
 
-    def get_or_create_image(self, url):
-        if url is None or len(url) == 0:
-            return None
-        if url in self._images:
-            return self._images[url]
-
-        defaults = {'publisher': self.organization, 'data_source': self.data_source}
-        img, created = Image.objects.get_or_create(
-            url=url, defaults=defaults)
-        return img
+    def set_image(self, obj, image_data):
+        if not image_data:
+            self._set_field(obj, 'image', None)
+            return
 
-    def set_image(self, obj, image_object):
-        if obj is None or image_object is None:
+        image_url = image_data.get('url', '').strip()
+        if not image_url:
+            print('Invalid image url "{}" obj {}'.format(image_data.get('url'), obj))
             return
-        if image_object not in obj.images.all():
+
+        image = self._get_image(image_url)
+        image = self._update_image(image, image_data)
+        self._set_field(obj, 'image', image)
+
+        if image._changed:
             obj._changed = True
-            obj.images.add(image_object)
+
+    def set_images(self, obj, images_data):
+        image_syncher = ModelSyncher(obj.images.all(), lambda image: image.url, delete_func=obj.images.remove)
+
+        for image_data in images_data:
+            image_url = image_data.get('url', '').strip()
+            if not image_url:
+                print('Invalid image url "{}" obj {}'.format(image_data.get('url'), obj))
+                continue
+
+            new_image = False
+            image = image_syncher.get(image_url)
+
+            if not image:
+                new_image = True
+                image = self._get_image(image_url)
+
+            image = self._update_image(image, image_data)
+
+            if new_image:
+                obj.images.add(image)
+                obj._changed = True
+            elif image._changed:
+                obj._changed = True
+
+            image_syncher.mark(image)
+
+        image_syncher.finish(force=True)
+
+    def _get_image(self, image_url):
+        if not image_url:
+            return None
+
+        if image_url in self._images:
+            return self._images[image_url]
+
+        image = Image(
+            publisher=self.organization,
+            data_source=self.data_source,
+            url=image_url,
+        )
+        image._changed = True
+        image._created = True
+
+        return image
+
+    def _update_image(self, image, image_data):
+        if not hasattr(image, '_changed'):
+            image._changed = False
+
+        self._set_field(image, 'publisher', self.organization)
```

Comment:
This line here reveals a bug in the underlying `_set_field` method (or actually the `Image` model `is_user_edited` method at https://github.com/City-of-Helsinki/linkedevents/blob/master/events/models.py#L144) that is only revealed when trying to import all tprek locations with the `tprek` importer. It fails with 

```
    return bool(self.data_source.user_editable and self.last_modified_by)
AttributeError: 'NoneType' object has no attribute 'user_editable'
```

This is because old existing images that we are updating do not necessarily have data source set, as they come from old times when `data_source` was missing from the image model.

This should be fixed by adding `bool(self.data_source)` to the Image model `is_user_editable` and `is_user_edited` methods.